### ISSUE TITLE
Pool seems to ignore concurrency setting

### ISSUE THREAD

**Issue by stefandoorn** [guzzle/guzzle#1506]
I'm trying to use a Pool, but for some reason (all I'm trying) the concurrency setting is getting ignored.

I'm doing:
- Building an array of Promises (asyncPost)
- When the console terminates (Symfony) I'm willing to fire them all but with some concurrency as it could be thousands

````
        // Build callable list of promises
        $requests = function ($command) {
            yield function() use ($command) {
                return $command->shiftOnePromise(); // Returns one Promise (Guzzlehttp\Promise\PromiseInterface)
            };
        };

        // Handle concurrent requests
        $pool = new Pool($client, $requests($command), [
            'concurrency' => 1,
            'fulfilled' => function ($response, $index) {
                // this is delivered each successful response
            },
            'rejected' => function ($reason, $index) {
                // this is delivered each failed request
            },
        ]);

        // Initiate the transfers and create a promise
        $promise = $pool->promise();

        // Force the pool of requests to complete.
        $promise->wait();
````

I have 25 promises to fulfil in this case, and it opens 25 connections, while I would expect that it stays with maximum of 1 connection and iterates over them. I'm not totally sure this is a bug/issue, but can't find another place to ask such a question.

Additional the shift promise function:

````
 /**
     * @return PromiseInterface
     */
    public function shiftOnePromise()
    {
        return array_shift($this->promises);
    }
````

**Comment by stefandoorn** [comment]
Did some additional digging, but it might be just my understanding of how the pool works and that I'm using it wrong. 

When creating the Pool object, the Iterable and the config are inserted and an EachPromise object is created after some additional work. After that, the `promise` method is called on the EachPromise object which calls the `createPromise` method. In this method itself, nowhere concurrency is checked but I have the impression all promises are handled in there and added to one big Promise, while for each promise the `wait` method is called (which fires them as far as I know). After that, the iterable is rewinded (in the `promise` method) and the `refillPending` method is called (in the `promise` method), in which concurrency is checked. But, there are no pending promises anymore, as they've already been processed in the `createPromise` method.

**Comment by kkopachev** [comment]
`createPromise` just creates `aggregate` and passes function. Aggregate promise will call this function only after `$eachPromise->wait()`. By that time `EachPromise` created aggregate (no execution of contained promises yet), according to concurrency limits added promises into `pending` array and advances `Iterable`.

**Comment by stefandoorn** [comment]
Alright. I probably got that wrong then. But, I'm seeing behavior I don't expect looking at the docs.

I'll try to show it with some code in here.

At several places in my code I register promises to an array holding them:

````
// Add request
        $promise = $this->client->requestAsync($type, $url, $params);
        $promise->then(
            function (ResponseInterface $response) use ($callback) {
                $this->output->writeln('Got response');
                if (is_callable($callback)) {
                    $callback($response);
                }
            },
            function (RequestException $e) {
                throw new \Exception(sprintf('HTTP error %s, reason: %s, response: %s',
                    $e->getResponse()->getStatusCode(), $e->getResponse()->getReasonPhrase(),
                    $e->getResponse()->getBody()->getContents()));
            }
        );
        $this->promises[] = $promise;
````

Then on termination of my console command I'm using an event listener to perform all these requests async:

````
        // Output
        $countPromises = $command->countPromises();
        if ($countPromises == 0) {
            return null;
        }
        $output->writeln(sprintf('<info>Starting processing of promises, count: %s, concurrency: %s</info>', $countPromises, self::CONCURRENCY));

        // Build callable list of promises
        $requests = function ($command) {
            yield function() use ($command) {
                /** @var Base $command */
                return $command->shiftOnePromise(); // This does array_shift one item of the array
            };
        };

        // Handle concurrent requests
        $pool = new Pool($command->getClient(), $requests($command), [
            'concurrency' => self::CONCURRENCY,
            'fulfilled' => function ($response, $index) {
                // this is delivered each successful response
            },
            'rejected' => function ($reason, $index) {
                // this is delivered each failed request
            },
        ]);

        // Initiate the transfers and create a promise
        $promise = $pool->promise();

        // Force the pool of requests to complete.
        $promise->wait();
````

I have an example running in here where I have 3 promises in the array, and concurrency set to 1. As soon as the code above fires, 3 connections to the client API are set up (one is re-used) directly and then 3 requests are performed directly. I expected to be using one connection and perform the second request only after the first one is finished, but all fire directly.

````
*   Trying {ip}... * About to connect() to {host} port 443 (#1)
*   Trying {ip}.... * About to connect() to  {host} port 443 (#2)
*   Trying {ip}.... * Connected to  {host} ({ip}) port 443 (#0)
````

The reason I'm willing to use queueing of promises in an array first is that I have two types of requests in my application: ones that I need the result of and ones that I don't need the result of. I first want to perform the ones I need the result of as they need to be synchronous all the time, and afterwards put the asynchronous ones in there fast.

**Comment by kkopachev** [comment]
Most likely your issue is that call to `$this->client->requestAsync($type, $url, $params);` actually sends request. Then it creates new promise for already finished response, on which you're attaching `then` callbacks. So currently your code is yielding already fulfilled promises to the pool. Then pool waiting on them one at a time.
To fix your code you should call `requestAsync` inside your generator, and yield them out.

**Comment by stefandoorn** [comment]
Yes, first I was thinking the same. But looking at the example here (http://docs.guzzlephp.org/en/latest/quickstart.html#concurrent-requests - last example + first example) I get the impression that it's ok to use the requestAsync function. But yes, it probably happens to start running already. So probably it's better to build an array of Request objects and iterate over them at the end of the script. 

Hopefully that way the callback functionality is also available them, as that was basically the main advantage I'm trying to achieve: set several different callbacks throughout code and run all promises at the end and fulfil the right callbacks.

In my believe the requestAsync didn't do anything until you call the wait() method on it, but apparently I'm wrong. Looking at the Client.php source code, I also have the impression that request() does fire wait() directly, but requestAsync does not. So, I'm a bit confused about this. I would expect requestAsync to not do anything until the wait() get's called.

**Comment by stefandoorn** [comment]
@mtdowling Could you also have a look on this issue and best practice? I'm a bit stuck in here and any help is appreciated :)

**Comment by kkopachev** [comment]
But you could always make an array of functions, which will return promises. 
```php
// Add request
$this->promisingFunctions[] = function () use($type, $url, $params, $callback) {
    $promise = $this->client->requestAsync($type, $url, $params);
    $promise->then(function (ResponseInterface $response) use ($callback) {
        $this->output->writeln('Got response');
        if (is_callable($callback)) {
            $callback($response);
        }
    }, function (RequestException $e) {
        throw new \Exception(sprintf('HTTP error %s, reason: %s, response: %s', $e->getResponse()->getStatusCode(), $e->getResponse()->getReasonPhrase(), $e->getResponse()->getBody()->getContents()));
    });

    return $promise;
}
```
and then just yield them in generator:
```php
$requests = function (Base $command) {
    yield $command->shiftOnePromise(); // This does array_shift one item of the array;
};
```

**Comment by mtdowling** [comment]
You need to yield promises from your generator or do as @kkopachev suggests, yield functions.

When you initiate an async transfer with requestAsync(), Guzzle will create a curl handle and add it to a shared curl multi instance. By queueing up a large list of promises, you're adding all of your promises to the multi handle at once, which means when you eventually call wait, you're waiting on all of the promises at once and not limiting your queue size at all.

**Comment by stefandoorn** [comment]
Thanks @kkopachev & @mtdowling, got it working! This also solved #1505, as they are also accurate now. Thanks a lot.
